chain/contracts/RoomStateMachine.sol

```diff
@@ -0,0 +1,268 @@
+    uint256 public seed;  // seed number for VRF generation.
+    uint256[6] public randomNumbers; // random numbers for card open.
+
+    constructor (
+        Params params,
+        bytes32 memory vrfPubKey,
+        bytes32 memory randSeedHash
+    ) payable {
+        params = params;
+        status = RoomStatus.WAIT_FOR_JOIN;
+        // bank is always the first to join
+        join(vrfPubKey, randSeedHash);
+    }
+
+    function join(bytes32 vrfPubKey, bytes32 randSeedHash) public payable {
+        require(status == RoomStatus.WAIT_FOR_JOIN, "invalid state machine status");
+        require(msg.value > params.minPlayerDeposit, "not enough deposit");
+        Player memory p = Player({
+            playerAddress: msg.sender,
+            vrfPubKey: vrfPubKey,
+            randSeedHash: randSeedHash,
+            balance: msg.value,
+            partialSeedRevealed: false,
+            vrfResultRevealed: false,
+            currentBet: 0,
+        });
+        players.push(p);
+        emit PlayerJoined(players.length);
+        if (players.length == params.playersNum) {
+            status = RoomStatus.WAIT_FOR_RANDOM_SEED
+        }
+    }
+
+    function numberOfPartialSeedsRevealed() private pure returns (uint) {
+        uint counter = 0;
+        for (int i=0; i<players.length; i++) {
+            if (players[i].partialSeedRevealed) {
+                counter++;
+            }
+        }
+        return counter;
+    }
+
+    function numberOfVRFResultRevealed() private pure returns (uint) {
+        uint counter = 0;
+        for (int i=0; i<players.length; i++) {
+            if (players[i].vrfResultRevealed) {
+                counter++;
+            }
+        }
+        return counter;
+    }
+
+    function numberOfBetPlaced() private pure returns (uint) {
+        uint counter = 0;
+        for (int i=0; i<players.length; i++) {
+            if (players[i].currentBet.option != BetOption.NONE) {
+                counter++;
+            }
+        }
+        return counter;
+    }
+
+    // game action
+    function reveal_seed(int playerIndex, uint256 randomNumber) private {
+        require(status == RoomStatus.WAIT_FOR_RANDOM_SEED, "invalid game status");
+        require(!players[playerIndex].partialSeedRevealed, "duplicated msg");
+        require(keccak256(randomNumber) == players[playerIndex].randSeedHash, "");
+        players[playerIndex].partialSeedRevealed = true;
+        seed = seed ^ randomNumber;  // merge partial seeds by xor together.
+        if (numberOfPartialSeedsRevealed() == params.playersNum) {
+            status = WAIT_FOR_BET;
+            expire_time = 0;  // recovery mode finished, cancel expire check.
+        }
+    }
+
+    // game action
+    function put_bet(int playerIndex, Bet bet) private {
+        require(status == RoomStatus.WAIT_FOR_BET, "invalid game status");
+        Player p = players[playerIndex];
+        require(p.currentBet.option == BetOption.NONE, "duplicated msg");
```

Comment:
When this happens, we can actually decide this player is cheating, and punish.
punishment is not implemented yet.

Reply:
it's find to not punish this action, let's keep it simple for now.